File: packages/connector-sdk/src/connector.ts
import { ActionDefinition, TriggerDefinition, AisDiscoveryResponse, AuthMethod, ConnectorCapabilities, ActionRequest, ActionResponse } from './types';

/**
 * AIS-compliant connector builder.
 */
export class AisConnector {
  private readonly name: string;
  private readonly version: string;
  private readonly actions: Map<string, ActionDefinition & { handler: Function }>;
  private readonly triggers: Map<string, TriggerDefinition>;
  private readonly authMethods: AuthMethod[];
  private readonly capabilities: ConnectorCapabilities;

  constructor(config: {
    name: string;
    version: string;
    actions?: (ActionDefinition & { handler: (input: any, ctx?: any) => Promise<any> })[];
    triggers?: TriggerDefinition[];
    authMethods?: AuthMethod[];
    capabilities?: Partial<ConnectorCapabilities>;
  }) {
    this.name = config.name;
    this.version = config.version;
    this.actions = new Map((config.actions || []).map((a) => [a.key, a]));
    this.triggers = new Map((config.triggers || []).map((t) => [t.key, t]));
    this.authMethods = config.authMethods || ['api_key'];
    this.capabilities = {
      read: true,
      write: true,
      webhook: false,
      batch: false,
      search: false,
      ...config.capabilities,
    };
  }

  /**
   * Get the AIS discovery response.
   * This is served at GET /.well-known/ais
   */
  getDiscovery(): AisDiscoveryResponse {
    return {
      aisVersion: '0.1',
      connectorName: this.name,
      connectorVersion: this.version,
      capabilities: this.capabilities,
      actions: Array.from(this.actions.values()).map(({ handler, ...action }) => action),
      triggers: Array.from(this.triggers.values()),
      authMethods: this.authMethods,
    };
  }

  /**
   * Execute a connector action by key.
   * This is served at POST /ais/actions/:key
   */
  async executeAction(req: ActionRequest): Promise<ActionResponse> {
    const start = Date.now();
    const action = this.actions.get(req.actionKey);

    if (!action) {
      return {
        success: false,
        error: `Unknown action: ${req.actionKey}`,
        meta: { durationMs: Date.now() - start, retryable: false },
      };
    }

    try {
      const data = await action.handler(req.input, req.context);
      return {
        success: true,
        data,
        meta: { durationMs: Date.now() - start, retryable: action.idempotent },
      };
    } catch (err) {
      return {
        success: false,
        error: err instanceof Error ? err.message : 'Unknown error',
        meta: { durationMs: Date.now() - start, retryable: action.idempotent },
      };
    }
  }

  /**
   * Get all action keys for validation purposes.
   */
  getActionKeys(): string[] {
    return Array.from(this.actions.keys());
  }

  /**
   * Validate input against an action's schema.
   */
  validateActionInput(actionKey: string, input: any): { valid: boolean; errors: string[] } {
    const action = this.actions.get(actionKey);
    if (!action) return { valid: false, errors: [`Unknown action: ${actionKey}`] };

    const errors: string[] = [];
    const schema = action.input;

    if (schema.required) {
      for (const field of schema.required) {
        if (input[field] === undefined || input[field] === null) {
          errors.push(`Missing required field: ${field}`);
        }
      }
    }

    return { valid: errors.length === 0, errors };
  }
}

File: packages/connector-sdk-python/src/aifut_connector_sdk/connector.py
"""
AIS-compliant connector builder — Python implementation.
"""

from typing import Any, Callable, Optional
from .types import ActionRequest, ActionResponse, AisDiscoveryResponse


class AisConnector:
    """
    AIS-compliant connector builder.

    Usage:
        connector = AisConnector(name="MyCRM", version="1.0.0", actions=[...])
        connector.serve(port=3000)  # Start FastAPI server
    """

    def __init__(
        self,
        name: str,
        version: str = "1.0.0",
        actions: Optional[list[dict]] = None,
        triggers: Optional[list[dict]] = None,
        auth_methods: Optional[list[str]] = None,
        capabilities: Optional[dict] = None,
    ):
        self.name = name
        self.version = version
        self._actions: dict[str, dict] = {}
        self._triggers: dict[str, dict] = {}
        self.auth_methods = auth_methods or ["api_key"]
        self.capabilities = {
            "read": True,
            "write": True,
            "webhook": False,
            "batch": False,
            "search": False,
            **(capabilities or {}),
        }

        if actions:
            for action in actions:
                key = action["key"]
                self._actions[key] = action

        if triggers:
            for trigger in triggers:
                key = trigger["key"]
                self._triggers[key] = trigger

    def get_discovery(self) -> dict:
        """Get the AIS discovery response (GET /.well-known/ais)."""
        return {
            "aisVersion": "0.1",
            "connectorName": self.name,
            "connectorVersion": self.version,
            "capabilities": self.capabilities,
            "actions": [
                {k: v for k, v in a.items() if k != "handler"}
                for a in self._actions.values()
            ],
            "triggers": list(self._triggers.values()),
            "authMethods": self.auth_methods,
        }

    async def execute_action(self, req: ActionRequest) -> ActionResponse:
        """Execute a connector action by key (POST /ais/actions/:key)."""
        import time
        start = time.time()

        action = self._actions.get(req.action_key)
        if not action:
            return ActionResponse(
                success=False,
                error=f"Unknown action: {req.action_key}",
                meta={"duration_ms": int((time.time() - start) * 1000), "retryable": False},
            )

        try:
            handler = action.get("handler")
            if handler is None:
                return ActionResponse(
                    success=False,
                    error=f"No handler for action: {req.action_key}",
                    meta={"duration_ms": int((time.time() - start) * 1000), "retryable": False},
                )

            # Call the handler (sync or async)
            result = handler(req.input, req.context)
            if hasattr(result, "__await__"):
                result = await result

            return ActionResponse(
                success=True,
                data=result,
                meta={
                    "duration_ms": int((time.time() - start) * 1000),
                    "retryable": action.get("idempotent", True),
                },
            )
        except Exception as e:
            return ActionResponse(
                success=False,
                error=str(e),
                meta={
                    "duration_ms": int((time.time() - start) * 1000),
                    "retryable": action.get("idempotent", True),
                },
            )

    def serve(self, host: str = "0.0.0.0", port: int = 3000):
        """Start a FastAPI server for the connector."""
        try:
            from fastapi import FastAPI, Request
            import uvicorn

            app = FastAPI(
                title=self.name,
                version=self.version,
                description="AIS-compliant connector",
            )

            @app.get("/.well-known/ais")
            async def discovery():
                return self.get_discovery()

            @app.post("/ais/actions/{action_key}")
            async def execute(action_key: str, request: Request):
                body = await request.json()
                req = ActionRequest(
                    action_key=action_key,
                    input=body.get("input", {}),
                    context={
                        "tenant_id": request.headers.get("x-tenant-id"),
                        "request_id": request.headers.get("x-request-id"),
                    },
                )
                result = await self.execute_action(req)
                status_code = 200 if result.success else 400
                return type('Response', (), {
                    'status_code': status_code,
                    'body': result.__dict__,
                    'json': lambda self: self.body,
                })()

            @app.get("/health")
            async def health():
                return {"status": "ok", "connector": self.name}

            print(f"Connector '{self.name}' serving at http://{host}:{port}")
            uvicorn.run(app, host=host, port=port)

        except ImportError:
            raise ImportError(
                "FastAPI is required to serve the connector. "
                "Install with: pip install aifut-connector-sdk[dev]"
            )
